EIPS/eip-2935.md

```diff
@@ -26,22 +27,34 @@ Additional secondary motivations include:
 
 | Parameter | Value |
 | - | - |
-| `FORK_BLKNUM` | TBD |
+| `FORK_TIMESTAMP` | TBD |
 | `HISTORY_STORAGE_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe`|
+| `HISTORY_SERVE_WINDOW` | `256` |
 
-At the start of processing any block where `block.number > FORK_BLKNUM` (ie. before processing any transactions), run `sstore(HISTORY_STORAGE_ADDRESS, block.number - 1, block.prevhash)`.
+At the start of processing any block where `block.timestamp > FORK_TIMESTAMP` (ie. before processing any transactions):
 
-When `block.number > FORK_BLKNUM + 256`, change the logic of the `BLOCKHASH` opcode as follows: if `FORK_BLKNUM <= arg < block.number`, return `sload(HISTORY_STORAGE_ADDRESS, arg)`. Otherwise return 0.
+ * run `sstore(HISTORY_STORAGE_ADDRESS, block.number - 1, block.prevhash)`. 
+ * **additionally** if this is the first block of the fork, then store last `min(HISTORY_SERVE_WINDOW-1, block.number-1)` **blocks' prevhash** into their corresponding slots i.e persist the entire `HISTORY_SERVE_WINDOW` block hashes (up until the genesis) that can be accessed in the block.
```

Comment:
```suggestion
 * **additionally** if this is the first block of the fork, then store last `min(HISTORY_SERVE_WINDOW, block.number)` **blocks' prevhash** into their corresponding slots i.e persist the entire `HISTORY_SERVE_WINDOW` block hashes (up until the genesis) that can be accessed in the block.
```

I think there is an off-by-one error here. If this EIP gets activated at block 0, then at block 1 we get `min(255, 0)` (thus 0) and therefore the "last 0 **blocks' prevhash**" get written to the contract. From other changes to the EIP I get the impression that we want the genesis hash to be written as well.

In general I get the idea that the idea of the BLOCKHASH opcode does not change during the 256 block-window of the fork happening, therefore I think we have this off-by-one-error.

In case of the fork happening at block 256: we would still have access to the `BLOCKHASH` of the blocks 0-255 (so 256 blocks in total), so we need to write the latest 256 blocks (so 0-255) to the contract, not the latest 255 blocks (so 1-255) (`min(256-1, 256-1) = min(255,255) = 255)` per original).

Reply:
this step is previous to additional step, so block 1 writes its prevhash i.e. 0 => genesis hash, and then no additional things get written

Reply:
Yup, I missed the `additional` point

Reply:
it's correct, because in the general case, you have to add 255 blocks after adding the one just before (notice the **additionally** in bold).

In the case of genesis, you also get `min(255, 0) == 0` so no extra block get inserted.

I think we should switch to a pseudo-python implementation, because you're the second person to be confused by this (I was the first :D)

Reply:
so something like:

```python3
def process_block_hash_history(block :Block, state: State):
    if block.timestamp >= FORK_TIMESTAMP:
        state.insert_slot(HISTORY_STORAGE_ADDRESS, block.number-1, block.parent.hash)

    # If this is the first fork block, add the parent's direct 255 ancestors as well
    if block.parent.timestamp < FORK_TIMESTAMP:
        ancestor = block.parent
        for i in range(255):
            ancestor = ancestor.parent
            state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number, ancestor.hash)
```